taichi/backends/metal/codegen_metal.cpp

```diff
@@ -1043,6 +1086,11 @@ class KernelCodegen : public IRVisitor {
     return kernel_name + "_func";
   }
 
+  void mark_print_used() {
+    TI_ASSERT(current_kernel_attribs_ != nullptr);
+    current_kernel_attribs_->uses_print = true;
```

Comment:
nit: Use a `UsedFeatures` structure like OpenGL does.

Reply:
Good point.. This is a sloppy design in Metal. I don't have a data structure that records the attributes of a **taichi** kernel. All i have is a vector of Metal kernels', see https://github.com/taichi-dev/taichi/blob/37976e750e1579ff9328f3d25f1a63cb3b3292e2/taichi/backends/metal/kernel_manager.h#L45.

I'd like to take this opportunity to clean up this flaw in another PR. Once I have a `TaichiKernelAttributes`, i can also add a `UsedFeatures` there. Then we don't even need to iterate over all the **metal** kernels to figure out if we need to flush the print buffers or not. (I tried adding something, but feel like it's too noisy for this one.) WDYT?

Reply:
Yes, ff2 do it iapr, you may also implement a `operator|=(UsedFeatures, UsedFeatures)` for recording all used feature among kernels :)